src/isc-dhcp/patch/0008-interface-name-maxlen-crash.patch

```diff
@@ -0,0 +1,13 @@
+diff --git a/common/discover.c b/common/discover.c
+index 3cd64a7..c85d18c 100644
+--- a/common/discover.c
++++ b/common/discover.c
+@@ -547,7 +547,7 @@ next_iface4(struct iface_info *info, int *err, struct iface_conf_list *ifaces) {
+ 				log_error("Interface name '%s' too long", name);
+ 				return 0;
+ 			}
+-			strncpy(info->name, name, sizeof(info->name) - 1);
++			strncpy(info->name, name, sizeof(info->name));
```

Comment:
@antony-rheneus: Can you please explain why you made this change? Subtracting 1 from the size of the buffer ensures that the copied string will be NULL-terminated. This change could potentially result in a non-terminated buffer. It seems like there is now the potential for a crash _with_ this change rather than without it.

Reply:
_Excerpt from Man page strncpy:-_ 
       The strncpy() function is similar, except that at most n bytes of src are copied.  Warning: If there is no null byte among the first n bytes of src, the string placed in dest  **will  not  be null-terminated.**

_Issue in Code:-_
Max Interface name size is 16 bytes.
info->name is NOT initialized/memset to 0 (NULL).
If source string "name" has 15 bytes like in case of "PortChannel0001", then earlier code will copy only 15 bytes (max_size - 1). So it would not copy the Last NULL character from the 16th byte of source string "name"
This leads to seg fault/crash


There is a check above this line that Source string "name" will not go beyond max size 16 bytes.
And Source string should be NULL terminated by the caller.
Now this strncpy line should ensure to add NULL terminator at the last byte. 

Your concern of Non-Terminated NULL buffer would arise only if source string "name" is NOT null terminated or is more than 15 bytes. But this would not happen as the below lines would catch it and return Error.

```
 546                         len = p - name;
 547                         if (len >= sizeof(info->name)) {

```

Reply:
@antony-rheneus: Thanks for the quick reply. I now understand your reasoning. However, a few lines below this, `name` is also copied into `tmp.ifr_name` in a similar fashion. However, in this case, the buffer _is_ initialized to 0:

```
    memset(&tmp, 0, sizeof(tmp));
    strncpy(tmp.ifr_name, name, sizeof(tmp.ifr_name) - 1);
```

Therefore, with your patch, these two copy operations differ, and at first glance it is not clear why. Thus, I think it makes more sense to keep these copy operations identical in behavior. So I suggest that we instead explicitly memset the buffer `info->name` to 0 before doing the strncpy.

```
    memset(&info->name, 0, sizeof(info->name));
    strncpy(info->name, name, sizeof(info->name) - 1);
```

I think this is the most easily understood code, and an explicit memset of `info->name` couldn't hurt, especially because they unnecessarily define it to be one byte longer than `IFNAMSIZ`.

```
struct iface_info {
    char name[IF_NAMESIZE+1];   /* name of the interface, e.g. "bge0" */
    struct sockaddr_storage addr;   /* address information */
    struct sockaddr_storage netmask;    /* netmask information */
    isc_uint64_t flags;     /* interface flags, e.g. IFF_LOOPBACK */
};
```

What are your thoughts on my proposal?

Reply:
@antony-rheneus In future, please don't include changes not related to main topic into a big PR like this. I approved at iteration 33, but you still made significant code changes after that. My approval is not applied to anything after that iteration.

I agree this `strncpy` is a valid bug fix. I notice another occurrence, please check whether it is a bug.
```c
		strncpy(info->name, name, sizeof(info->name) - 1);

#ifdef SKIP_DUMMY_INTERFACES
	} while (strncmp(info->name, "dummy", 5) == 0);
#else
	} while (0);
#endif
```

Reply:
@qiluo-msft: Where is the occurrence you refer to? Is it the line I mentioned above, or something else? There is no second reference in the snippet you posted.

Reply:
@jleveque ,
From which file you referred iface_info? I could see below in SONIC dhcp submodule for name[IFNAMSIZ]
```
src/isc-dhcp/isc-dhcp/common/discover.c
 399 /*
 400  * Structure used to return information about a specific interface.
 401  */
 402 struct iface_info {
 403         char name[IFNAMSIZ];            /* name of the interface, e.g. "eth0" */
 404         struct sockaddr_storage addr;   /* address information */
 405     struct sockaddr_storage netmask;    /* netmask information */
 406         isc_uint64_t flags;             /* interface flags, e.g. IFF_LOOPBACK */
 407 };
 408
```

I agree with your proposal, by doing memset before strncpy.
However using strncpy in right way will avoid cpu cycles of memset operation.

Reply:
@antony-rheneus: I mistakenly quoted from the Solaris section at lines 227-230. The definition you quoted is the Linux definition, so the buffer is only `IFNAMSIZ` is length.

I'm not very concerned about a few extra CPU cycles here, as interface discovery only occurs once when the relay agent starts.

Also, I respectfully disagree that your solution is the "right way" to use strncpy. This is debatable. I personally think the proper way to use strncpy is to explicitly write a NUL char after the strncpy call. This not only ensures there will be no overflow, but it is easy to read and understand what is taking place. In the case of this code, a memset first will accomplish the same thing, but if I wrote the code myself, I would have explicitly written a NUL char after the strncpy call.

Reply:
I agree.

Reply:
FYI, The following upstream patch introduced a memset call before the strncpy call, much like I had suggested earlier in these comments: https://gitlab.isc.org/isc-projects/dhcp/-/commit/b7bfb2c74821ad6055c349dd2eddf2c749578762. This patch was included in the 4.4.0 release of isc-dhcp.

We are [upgrading](https://github.com/Azure/sonic-buildimage/pull/4671) the dhcp_relay Docker container to be based off Debian Buster, for which the native version of isc-dhcp is currently 4.4.1-2. Therefore, I have requested this patch be removed as part of the upgrade, as it is no longer necessary.